### ISSUE TITLE
Implement Auto Search

### ISSUE THREAD

**Issue by grumpyinca** [thegrumpys/odop#880]
### Summary:
The notion of automatically triggering the ODOP search feature has come up for discussion many times.  A design article and other issues have already developed various aspects of an implementation.  These include: 
 - [Auto Search Design article](https://odop.herokuapp.com/docs/design/AutoSearch.html) 
 - #25 
 - #221 
 - #583 
 - #801

### Other aspects of the implementation:
 - Currently, character-by-character changes to Independent Variables cause recalculation and display updates. Using the same approach with Auto Search may create a performance issue.  Worse, Auto Search on a partially formed input will likely cause the current design to move to remote parts of the design space potentially introducing a Search reliability problem.  An alternative input approach, such as conducting the search only after the input field loses focus, or on input of enter or tab keys, may be desirable.
 - Significant performance improvements or spinner display while searching should be in place before an Auto Search feature is released. 
 - A Preference setting to disable Auto Search behavior and revert to current manual calculation is desired. 
 - Should Auto Search apply only when Fixing Dependent Variables ?
 - Should Auto Search apply only in Calculator View ?

**Comment by grumpyinca** [comment]
Feedback from review of first prototype

1. Auto Fix & Auto Search checkbox fields overlap.
2. Search will run if design is already feasible.  This is not desirable because it produces the " ... may produce low precision results" warning.
3. Set focus, then unset focus produces a Search.  The user may have just been using the cursor as a place holder for further thought and not expected the numbers to change on moving the cursor elsewhere. 
4. Get stale, unexpected results after entering an invalid (red) value then changing focus. 
5. With Auto fix disabled:  Set value. Click Fix.  Search is performed as Not Fixed.  (Not sure that there is an easy fix for this one!). 
6. It may not be repeatable, but I seemed to get inconsistent behavior in making changes to Force_2 as compared to similar changes in Force_1.

**Comment by grumpyinca** [comment]
### Continuing testing of first prototype

When importing with the shortcut to a previously imported design: 

```
Uncaught runtime errors:
×
ERROR
Cannot read properties of null (reading 'type')
TypeError: Cannot read properties of null (reading 'type')
    at FileReader.onLoadEnd (http://localhost:3000/static/js/bundle.js:76869:105)
```
I do not yet have the steps to reliably reproduce this.

**Comment by grumpyinca** [comment]
### Auto Search UI decisions / priorities

1. Search should run after a checkbox loses focus ... If the design is not feasible and if the checkbox is a fix transitioning to free or if the checkbox is a constraint being disabled.  The current implementation where search immediately runs on transition from fixed to free without the checkbox losing focus is over-kill and inconsistent with my latest visualization of desired behavior.

2. Action:Select Size does not trigger Auto Search.  I think that it should.

3. Action:Select Catalog does not trigger Auto Search.  I think that it should not.

4. What about arrow keys ?  I am OK with ignoring them.

5. Do we have any kind of Validity check operating before Auto Search ?

6.  In Calculator View, Auto Search will run while the Edit Independent Variable dialog box is still up.  While this is different from my initial thoughts about the desired behavior, I think that I can live with it.

**Comment by 1fifoto** [comment]
* After changing NVUIV to onSetFix & onResetFix, in SymbolValue and SymbolValueWireDia do the same for NVUDV, and their constraints. SymbolValueWireDia has its own copy of NVUIV - ugh!
* Change SymbolValueWireDia when changing to use standard wire sizes (see Action Select Size). Now they do it differently. Mike prefers Action Select Size to do its function similar to SymbolValueWireDia.

**Comment by grumpyinca** [comment]
### Proposed text of Auto Search "i" in a circle.

                  <p>When checked, enables Auto Search.</p>
                  <p>When enabled, The Search feature executes when changes are made and the current design is not feasible.
                         Search is triggered when the changed field "loses focus".  
                         Specifically, when the cursor moves away from the changed field.</p>
                  <p>The behavior is the same as the File : Preferences enable_auto_search value.</p>

**Comment by 1fifoto** [comment]
Proposed text adjusted. What do you think?

                  <p>When checked, enables Auto Search.</p>
                  <p>When Auto Search is enabled, a Search is triggered when the current design is not feasible and the changed field "loses focus".  
                     Specifically, when the cursor focuses on some other field.</p>
                  <p>The behavior is the same as the File : Preferences enable_auto_search value.</p>

**Comment by grumpyinca** [comment]
I like your improvement(s).  Here is a slightly tweaked counter proposal:

              <p>When checked, enables Auto Search.</p>
              <p>When Auto Search is enabled, a Search is automatically triggered when the current design is not feasible 
                     and the changed field "loses focus".    
                     Specifically, when the cursor moves away from the changed field and focuses on some other field or background. 
              </p>
              <p>When Auto Search is not enabled, The Search feature executes only when a Search (solve) button is pushed or the 
                      Action : Search (solve) menu entry is invoked.
              <p>The behavior is the same as the File : Preferences enable_auto_search value.</p>

**Comment by 1fifoto** [comment]
There is a stale data problem in actionSelectSize's onSelect when auto search is enabled. The onSelect function updates the model with a new wire-dia or od_free using changeSymbolValue(name, value) and then it checks if a auto search should be done by checking of model_enable_auto_search is true, the value has changed, and the model_objective_value >= model_objmin. The first two conditions are true, but the third "model_objective_value >= model_objmin" is stale and does NOT represent the current state of the model after the value was changes. As a result, the search is not done when it should have been executed.

**Comment by grumpyinca** [comment]
### Current version test results

 1. Not getting log_usage records for auto searches.  
 2.

**Comment by 1fifoto** [comment]
Note: Auto Search is implemented in the UI's Focus/Blur/KeyPress handlers and not in the Redux model. As a result Execute/Demo/Tutorials do not use the UI, but instead use the Redux model. Therefore, and like Auto Fix, the Auto Search is not affected by nor does it affect the Execute/Demo/Tutorials.

**Comment by grumpyinca** [comment]
### Repeatable case that unexpectedly invokes Seek
Steps to observe the problem:

1. Restart system; confirm that developer console is open
2. Load extension spring (continue past migrate prompt)
3. Switch to Calculator View (not able to duplicate in Advanced View)
4. Click on Stress_Initial   
   - Set MIN Constraint to a constant value (I use 10,000)
   - Click Close
5. Click on Force_2
   - Enable MIN Constraint 
   - Set MIN constraint value to 42
   - Click Close button  **<--- This triggers the bug**

**Notes**:

1. Not able to reproduce the problem in Advanced View.
2. While it appears to be necessary to change an FDCL variable to Non-FDCL, it does not appear to matter which one. 
3. You must click on the Close button to trigger the problem.  Just losing focus is not enough.
4. The close button is suspiciously close to the position of the Seek MAX Force_2 button before the Force_2 value change to 42.

**Comment by grumpyinca** [comment]
Shortened & simplified the AlertsAcordian.jsx text describing the Auto Search feature to: 
```
                  <p>When checked, a Search is automatically triggered when the current design is not feasible
                     and the cursor moves away from a changed value.
                     For more detail, enter "AutoSearch" in Help Lookup.</p>
                  <p>When unchecked, the Search feature executes only when a Search (solve) button is pushed or the
                     Action : Search (solve) menu entry is invoked.</p>
                  <p>The behavior is the same as the File : Preferences enable_auto_search value.</p> 
```

Added new online Help topic titled "Auto Fix, Auto Search"

**Comment by grumpyinca** [comment]
### Proposed text to be added to the first page of selected demo & tutorial sessions

```
When operating in the demo / tutorial environment (text with light green background),
operation of various features including Search, Seek and Trade is controlled by the tutorial.
Changing the Auto Fix and Auto Search check boxes will not impact the outcome of individual demo and tutorial sessions. 
Regardless of the current visual status of those check boxes, 
the demo and tutorial sessions operate as though they are not checked. 
```

**Comment by grumpyinca** [comment]
When Auto Search is enabled, should Search run after **Action : Execute** ?

Would it be difficult for a script (increasePrecision is a good candidate) to leave a flag triggering a Search?

**Comment by grumpyinca** [comment]
Reset after Auto Search seems to work well.

Question:  In Advanced view, would it make any sense for the escape key to work like the Reset button does in Calculator View ?

**Comment by 1fifoto** [comment]
Merged into master, closing
